libs/orderbook/src/lib/orderbook-data.ts

```diff
@@ -0,0 +1,310 @@
+import produce from 'immer';
+import groupBy from 'lodash/groupBy';
+import type {
+  MarketDepth_market_depth_sell,
+  MarketDepth_market_depth_buy,
+} from './__generated__/MarketDepth';
+import type {
+  MarketDepthSubscription_marketDepthUpdate_sell,
+  MarketDepthSubscription_marketDepthUpdate_buy,
+} from './__generated__/MarketDepthSubscription';
+
+export interface CumulativeVol {
+  bid: number;
+  relativeBid?: number;
+  ask: number;
+  relativeAsk?: number;
+}
+
+export interface OrderbookData {
+  price: number;
+  bidVol: number;
+  bidVolByLevel?: Record<number, number>;
+  relativeBidVol?: number;
+  askVol: number;
+  askVolByLevel?: Record<number, number>;
+  relativeAskVol?: number;
+  cumulativeVol: CumulativeVol;
+}
+
+const getGroupPrice = (price: number, resolution: number) =>
+  Math.round(price / resolution) * resolution;
+
+const maxVolumes = (orderbookData: OrderbookData[]) => {
+  let bidVol = 0;
+  let askVol = 0;
+  let cumulativeVol = 0;
+  orderbookData.forEach((data) => {
+    bidVol = Math.max(bidVol, data.bidVol);
+    askVol = Math.max(askVol, data.askVol);
+  });
+  cumulativeVol = Math.max(
+    orderbookData[0]?.cumulativeVol.ask,
+    orderbookData[orderbookData.length - 1]?.cumulativeVol.bid
+  );
+  return {
+    bidVol,
+    askVol,
+    cumulativeVol,
+  };
+};
+
+const toPercentValue = (value?: number) => Math.ceil((value ?? 0) * 100);
+
+const updateRelativeData = (data: OrderbookData[]) => {
+  const { bidVol, askVol, cumulativeVol } = maxVolumes(data);
+  data.forEach((data) => {
+    data.relativeAskVol = toPercentValue(data.askVol / askVol);
+    data.relativeBidVol = toPercentValue(data.bidVol / bidVol);
+    data.cumulativeVol.relativeAsk = toPercentValue(
+      data.cumulativeVol.ask / cumulativeVol
+    );
+    data.cumulativeVol.relativeBid = toPercentValue(
+      data.cumulativeVol.bid / cumulativeVol
+    );
+  });
+};
+
+export const compact = (
+  sell:
+    | (
+        | MarketDepth_market_depth_sell
+        | MarketDepthSubscription_marketDepthUpdate_sell
+      )[]
+    | null,
+  buy:
+    | (
+        | MarketDepth_market_depth_buy
+        | MarketDepthSubscription_marketDepthUpdate_buy
+      )[]
+    | null,
+  resolution: number
+) => {
+  let cumulativeVol = 0;
+  const askOrderbookData = [...(sell ?? [])]
+    .sort((a, b) => Number(a.price) - Number(b.price))
+    .map<OrderbookData>((sell) => ({
+      price: Number(sell.price),
```

Comment:
suggestion (blocking, but could also potentially be some follow up work): Lots of math going on here with numbers that potentially have lots of decimal places. We should probably use BigNumber, this will have the knock on effect of having to use BigNumbers methods like .plus rather than regular JS operations.

```suggestion
      price: new BigNumber(sell.price),
```

Reply:
I changed price type to string and I use BigNumber(price) for to compare which one is greater than the other